Yt/ch13313/rle for strings

ypatia opened the pull request:
This PR is solely about adding a compressor for RLE encoding of strings. Currently RLE filter only support integers. This story is just adding the encoding classes, not plumbing them into the filter pipeline. As such it is not taking into account offsets and other particularities of var_sized attributes.

Note: I am using [span](https://en.cppreference.com/w/cpp/container/span), a C++20 standardized construct, so I am bringing in a header file with its implementation, like it was done in TileDB-VCF. I need some CMake help to put this header in the right place (I imagine in external directory) as I tried and I failed.


---
TYPE: IMPROVEMENT
DESC: Adding a compressor algorithm for RLE encoding of strings

eric-hughes-tiledb reviewed (changes requested):
If we're going to add <span>, we should do it for the whole code base, not just a single module. I've just today been looking at using `gsl::span` for this, FYI. In addition, we've already got three pieces of outside code copied into our source tree (i.e. not external CMake dependencies) in `/external`, so any code we'd add like this should go there.

Regardless, I'd really prefer to add <span> in a separate PR where we can discuss selection of implementation without turning this PR into that discussion.

ypatia commented:
I agree about adding it in the right folder, that is why I wrote in this PR description (maybe I should have added it as a comment to make it more visible in the first place) : 
> Note: I am using span, a C++20 standardized construct, so I am bringing in a header file with its implementation, like it was done in TileDB-VCF. I need some CMake help to put this header in the right place (I imagine in external directory) as I tried and I failed.

No problem with me if you'd like to introduce `span` in a separate PR though, I just brought in the .hpp that is currently used in TileDB-VCF. I'll let @Shelnutt2 comment on how urgent this current PR is and if it can wait for another PR for span as you are suggesting.

ypatia commented:
> @eric-hughes-tiledb I agree about adding it in the right folder, that is why I wrote in this PR description (maybe I should have added it as a comment to make it more visible in the first place) :
> 
> > Note: I am using span, a C++20 standardized construct, so I am bringing in a header file with its implementation, like it was done in TileDB-VCF. I need some CMake help to put this header in the right place (I imagine in external directory) as I tried and I failed.
> 
> No problem with me if you'd like to introduce `span` in a separate PR though, I just brought in the .hpp that is currently used in TileDB-VCF. I'll let @Shelnutt2 comment on how urgent this current PR is and if it can wait for another PR for span as you are suggesting.

Replaced `span` implementation with the one chosen ( https://github.com/tcbrindle/span ) 
Added the header file in `external` and alias in `common.h`

ypatia commented on tiledb/common/common.h:
```diff
@@ -56,4 +56,7 @@ using tiledb::common::allocator;
 
+#include "external/include/span/span.hpp"
+using tcb::span;
+
```
fixed

ypatia replied:
FYI there is an issue on windows when bringing `std::byte` into the global namespace, as it clashes with a `byte` definition in a windows header file (`rpcndr.h`). More details here: https://developercommunity.visualstudio.com/t/error-c2872-byte-ambiguous-symbol/93889

To work around this issue I reverted the type aliasing for byte for now, and went back to use `std::byte`.

eric-hughes-tiledb commented on tiledb/sm/compressors/rle_compressor.h:
```diff
@@ -73,8 +78,104 @@ class RLE {
+   * @tparam T Type of integer to store run legths, must fit max num of
+   * repetitions in input
+   * @tparam P Type of integer to store string sizes, must fit max length input
+   * string
```
C++ specifically says that `size_t` is a type that can represent the size of the largest objects that can fit in memory. It seems that these type parameters are extraneous at best.

Similarly, using `uint64_t` for byte sizes elsewhere isn't correct, strictly speaking. This is nowhere near the only place where this issue arises, but we don't need to be making the problem worse.

We do need to use `uint64_t` for lengths that don't need to fit into memory, notably file sizes. This is all to say that we don't have a good way of dealing with this issue yet. We need at least another size type defined (perhaps more than one) and a convention for dealing with coherently.

eric-hughes-tiledb commented on tiledb/sm/compressors/rle_compressor.h:
```diff
@@ -73,8 +78,104 @@ class RLE {
+    T run_length = 1;
+    auto out_offset = &output[0];
+    auto previous = input[0];
+    for (uint64_t in_index = 1; in_index < input.size(); in_index++) {
```
Using `decltype(input.size()) in_index` means that you don't need to worry about what size type `size()` returns.

eric-hughes-tiledb commented on tiledb/sm/misc/endian.h:
```diff
@@ -57,6 +61,23 @@ inline bool is_big_endian() {
+ * Compile-time endianness swap based on
+ * https://en.cppreference.com/w/cpp/language/fold
+ */
+template <class T, std::size_t... N>
```
If we need `using std::size_t` in our common header, we should add it.

eric-hughes-tiledb reviewed (changes requested):
I'm not that concerned with fixing all our size type problems in this PR.

The random test data do need addressing, though.

Shelnutt2 commented:
@ypatia can you rebase to get the fix for the GCS failure?
